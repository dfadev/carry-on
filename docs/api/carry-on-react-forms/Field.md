---
id: Field
title: Field
---
<html><head /><body /></html>