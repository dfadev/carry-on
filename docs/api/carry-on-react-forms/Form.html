<!DOCTYPE html><html lang="en"><head><meta charSet="utf-8"/><meta http-equiv="X-UA-Compatible" content="IE=edge"/><title>Form · carry-on</title><meta name="viewport" content="width=device-width"/><meta name="generator" content="Docusaurus"/><meta name="description" content="State container for React"/><meta name="docsearch:language" content="en"/><meta property="og:title" content="Form · carry-on"/><meta property="og:type" content="website"/><meta property="og:url" content="https://dfadev.github.io//carry-on/index.html"/><meta property="og:description" content="State container for React"/><meta property="og:image" content="https://dfadev.github.io//carry-on/img/docusaurus.png"/><meta name="twitter:card" content="summary"/><meta name="twitter:image" content="https://dfadev.github.io//carry-on/img/docusaurus.png"/><link rel="shortcut icon" href="/carry-on/img/favicon.ico"/><link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css"/><link rel="stylesheet" href="/carry-on/css/main.css"/></head><body class="sideNavVisible separateOnPageNav"><div class="fixedHeaderContainer"><div class="headerWrapper wrapper"><header><a href="/carry-on/"><img class="logo" src="/carry-on/img/card-travel.svg" alt="carry-on"/><h2 class="headerTitleWithLogo">carry-on</h2></a><div class="navigationWrapper navigationSlider"><nav class="slidingNav"><ul class="nav-site nav-site-internal"><li class="siteNavGroupActive"><a href="/carry-on/docs/" target="_self">Overview</a></li><li class="siteNavGroupActive"><a href="/carry-on/docs/api/" target="_self">API</a></li><li class=""><a href="https://github.com/dfadev/carry-on" target="_self">GitHub</a></li></ul></nav></div></header></div></div><div class="navPusher"><div class="docMainWrapper wrapper"><div class="container docsNavContainer" id="docsNav"><nav class="toc"><div class="toggleNav"><section class="navWrapper wrapper"><div class="navBreadcrumb wrapper"><div class="navToggle" id="navToggler"><i></i></div><h2><i>›</i><span>API</span></h2><div class="tocToggler" id="tocToggler"><i class="icon-toc"></i></div></div><div class="navGroups"><div class="navGroup"><h3 class="navGroupCategoryTitle">Overview</h3><ul><li class="navListItem"><a class="navItem" href="/carry-on/docs/">Getting Started</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/registeringState">Registering State</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/usingState">Using State</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/examples/">Examples</a></li></ul></div><div class="navGroup"><h3 class="navGroupCategoryTitle">API</h3><ul><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/">Info</a></li><div class="navGroup subNavGroup"><h4 class="navGroupSubcategoryTitle">carry-on-store</h4><ul><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-store/deleteStore">deleteStore</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-store/devTools">devTools</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-store/initStores">initStores</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-store/register">register</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-store/subscribe">subscribe</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-store/useStore">useStore</a></li></ul></div><div class="navGroup subNavGroup"><h4 class="navGroupSubcategoryTitle">carry-on-react</h4><ul><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-react/State">&lt;State ...&gt;</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-react/withState">withState</a></li></ul></div><div class="navGroup subNavGroup"><h4 class="navGroupSubcategoryTitle">carry-on-react-forms</h4><ul><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-react-forms/Field">Field</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-react-forms/form">form</a></li><li class="navListItem navListItemActive"><a class="navItem" href="/carry-on/docs/api/carry-on-react-forms/Form">Form</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-react-forms/FormButtons">FormButtons</a></li><li class="navListItem"><a class="navItem" href="/carry-on/docs/api/carry-on-react-forms/FormState">FormState</a></li></ul></div></ul></div></div></section></div><script>
            document.addEventListener('DOMContentLoaded', function() {
              createToggler('#navToggler', '#docsNav', 'docsSliderActive');
              createToggler('#tocToggler', 'body', 'tocActive');

              const headings = document.querySelector('.toc-headings');
              headings && headings.addEventListener('click', function(event) {
                if (event.target.tagName === 'A') {
                  document.body.classList.remove('tocActive');
                }
              }, false);

              function createToggler(togglerSelector, targetSelector, className) {
                var toggler = document.querySelector(togglerSelector);
                var target = document.querySelector(targetSelector);

                if (!toggler) {
                  return;
                }

                toggler.onclick = function(event) {
                  event.preventDefault();

                  target.classList.toggle(className);
                };
              }
            });
        </script></nav></div><div class="container mainContainer"><div class="wrapper"><div class="post"><header class="postHeader"><h1 class="postHeaderTitle">Form</h1></header><article><div><span></span></div></article></div><div class="docs-prevnext"><a class="docs-prev button" href="/carry-on/docs/api/carry-on-react-forms/form"><span class="arrow-prev">← </span><span>form</span></a><a class="docs-next button" href="/carry-on/docs/api/carry-on-react-forms/FormButtons"><span class="function-name-prevnext">FormButtons</span><span class="arrow-next"> →</span></a></div></div></div><nav class="onPageNav"></nav></div><footer class="nav-footer" id="footer"><section class="copyright">Copyright © 2019 Russ Panula</section></footer></div></body></html>