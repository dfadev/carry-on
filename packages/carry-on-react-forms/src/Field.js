/** @format **/
import React from "react";
import { State } from "carry-on-react";
import { getIn } from "carry-on-utils";

function getVal({ target: { type, value, checked } }) {
  if (type === "number" || type === "range") {
    const parsed = parseFloat(value);
    return Number.isNaN(parsed) ? "" : parsed;
  }

  if (type === "checkbox" || type === "radio") return checked;

  return value;
}

export default ({
  store,
  form = "form",
  path,
  default: def,
  children,
  type,
  ...rest
}) => (
  <State path={form} from={store} {...rest}>
    {({
      values,
      touched,
      errors,
      setFieldValue,
      isTouched,
      setFieldTouched,
      setFieldError
    }) =>
      children({
        touched: getIn(touched, path, false),
        error: getIn(errors, path, undefined),
        element: {
          onChange: e => setFieldValue(path, getVal(e)),
          onBlur: () => !isTouched(path) && setFieldTouched(path, true),
          [type === "checkbox" || type === "radio"
            ? "checked"
            : "value"]: getIn(values, path, def)
        },
        setValue: val => setFieldValue(path, val),
        setTouched: val => setFieldTouched(path, val),
        setError: val => setFieldError(path, val)
      })
    }
  </State>
);
