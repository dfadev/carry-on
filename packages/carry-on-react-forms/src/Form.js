/** @format **/
import React from "react";
import { State } from "carry-on-react";

export default ({ store, form = "form", children, ...rest }) => (
  <State path={form} from={store} {...rest}>
    {state => (
      <form onSubmit={state.submit} onReset={state.reset}>
        {children}
      </form>
    )}
  </State>
);
