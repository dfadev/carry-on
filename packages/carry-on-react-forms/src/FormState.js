/** @format **/
import React from "react";
import { State } from "carry-on-react";

export default ({ select, store, form = "form", children }) => (
  <State
    from={store}
    select={state => (select ? select(state[form]) : state[form])}
  >
    {children}
  </State>
);
