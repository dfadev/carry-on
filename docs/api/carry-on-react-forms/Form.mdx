---
id: Form
title: Form
---
import BrowserOnly from "@docusaurus/BrowserOnly";
import Inspector from "react-inspector";
import { FormState } from "carry-on-react-forms";
import { State } from "carry-on-react";
import theme from "../../inspector-theme";

## Import

```js

import { Form } from "carry-on-react-forms";

```

## `Form`


```js live noInline
State.Debug = true;
State.Verbose = true;

const formStateFactory = (formId = "form") => ({
  state: ({ get, set }) => ({
    [formId]: {
      initialValues: {
        field1: "value1",
        field2: "value2"
      },
      onSubmit(vals) {
        alert(JSON.stringify(vals));
      }
    }
  })
});

render(
  <Form id="myForm" register={formStateFactory("myForm")}>
    <div>
      <Field path="field1">
        {field => <input {...field.element} />}
      </Field>
    </div>
    <div>
      <Field path="field2">
        {field => <input {...field.element} />}
      </Field>
    </div>
    <div>
      <button type="submit">OK</button>
    </div>
  </Form>
);
```

<BrowserOnly>
  {() => {
    <State select={s => ({ ...s })}>
      {state => <Inspector data={state} theme={theme} expandLevel={2} />}
    </State>
  }}
</BrowserOnly>
