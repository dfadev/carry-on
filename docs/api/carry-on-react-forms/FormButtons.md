---
id: FormButtons
title: FormButtons
---
## Import

```js

import { FormButtons } from "carry-on-react-forms";

```

## `FormButtons`


```js


```
