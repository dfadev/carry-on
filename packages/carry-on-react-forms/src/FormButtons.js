/* eslint-disable react/jsx-curly-newline */
import React from "react";
import FormState from "./FormState";

export default function FormButtons({ children = () => null, ...rest }) {
  return (
    <FormState {...rest}>
      {({ submit, reset, isPristine, isValidating, isValid } = {}) =>
        children({
          submit: {
            onClick: submit,
            disabled: isPristine || isValidating || !isValid
          },
          reset: {
            onClick: reset,
            disabled: isPristine || isValidating
          }
        })
      }
    </FormState>
  );
}
