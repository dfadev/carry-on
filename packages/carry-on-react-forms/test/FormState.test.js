import React from "react";
import { render } from "react-testing-library";
import Form from "../src/Form";
import Field from "../src/Field";
import FormButtons from "../src/FormButtons";
import FormState from "../src/FormState";

it("should render <FormState>", () => {
  expect(render(<FormState />).asFragment()).toMatchSnapshot();
});
