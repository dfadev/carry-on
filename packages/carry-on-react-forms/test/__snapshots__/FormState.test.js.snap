// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render <FormState> 1`] = `<DocumentFragment />`;
